### ISSUE TITLE
Insecure random_int for PHP5.x

### ISSUE THREAD

**Issue by Sopsy** [postfixadmin/postfixadmin#182]
I hope this is the correct place to discuss this.

https://github.com/postfixadmin/postfixadmin/blob/6100ca4cf939968638fe5b8b139051a980438b55/functions.inc.php#L837

Is there a reason mt_rand() is used here as the first fallback? This is horribly insecure, we're talking about password hashing after all.
I propose that we should check if openssl_random_pseudo_bytes() exists and use that if it does, the bytes just need to be converted to numbers.

Better yet, bump the required PHP version to PHP7, that would also allow for a way better codebase. That would mean security releases for 3.x branch, and then a 4.x branch for PHP7. Have you had any plans to do this?

**Comment by DavidGoodwin** [comment]
It should be the case that random_compat is in use, in which case the random_int() function in functions.inc.php isn't used.

But obviously an end user might not have Phar loaded; or they might be on a crap version of PHP.

I don't think it's quite feasible to turn off PHP5 support - as there are a number of LTS distros which ship/support PHP5. 

So ... perhaps we'd be best using the random_compat code inline (rather than the bundled phar file)?

I think we're best not trying to maintain our own random_int() function really...

**Comment by Sopsy** [comment]
>I don't think it's quite feasible to turn off PHP5 support - as there are a number of LTS distros which ship/support PHP5.
That's true, yeah. Unfortunately.

One can generate a random int with openssl e.g. by doing this:
`hexdec(bin2hex(openssl_random_pseudo_bytes(2)))`
But then again the exact range cannot be defined, for 1 byte is 0-255, for 2 bytes is 0-65535 and so on, so the code using the function would need to be changed to account for this, so it's not that easy. And maybe just using the HEX string is not secure enough for a salt.

>I think we're best not trying to maintain our own random_int() function really...
Yeah, that's not good. Doing your own crypto should always be the last resort.

>So ... perhaps we'd be best using the random_compat code inline (rather than the bundled phar file)?
I would say leave it to the user to install it and then point to it in the configuration and disallow using any password functions completely if it does not exist.

While I understand the fallback and while not exactly as insecure as the comparison, but I would still compare it to storing passwords plaintext because the required cryptographic extensions do not exist.

**So in short:** Do not create an insecure fallback, instead add random_compat as a dependency for older PHP versions and nag the user (die()) when it does not exist.  Everyone who is able to install postfixadmin should be able to install it as well.

What do you think?

**Comment by DavidGoodwin** [comment]
https://github.com/postfixadmin/postfixadmin/commit/b6f75c9c2c48c22000a95b3a908c7c63d5d37be1

https://github.com/postfixadmin/postfixadmin/commit/318ac048d529b4ac21eae949a3a0540b89bed8af#diff-b6f76336fbaec4b0bae7029939f8fce0R827

(Sorry; it seems I failed to split my commits up in teh second bit).

**Comment by Sopsy** [comment]
That's awesome! Thanks for the proper fix!

**Comment by marcovtwout** [comment]
Aside question: is postfixadmin fully PHP7-compatible? Couldn't find it in the changelog.

**Comment by DavidGoodwin** [comment]
PHP-7 compatible - yes ... 

Note it's still compatible with PHP5.2+ - so we still use array(...) rather than [...]. ... but it should run without problem in PHP7, and if it doesn't file a ticket! :)

**Comment by Sopsy** [comment]
I actually run it on PHP7.2 without any issues.
